multi-component-workload tutorial creation (#29)
karmada-io/playground: karmada-multi-component-workload-example/step3/text.md

### Initialize Karmada control plane

Multi-component scheduling (`MultiplePodTemplatesScheduling`) is currently an **Alpha** feature in Karmada and is **disabled by default**. We can enable it directly at initialization time using the `--karmada-*-extra-args` flags, which pass additional command-line arguments to the respective control plane components:

RUN `karmadactl init --karmada-controller-manager-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true" --karmada-scheduler-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true" --karmada-webhook-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true"`{{exec}}

> This sets up the Karmada control plane on the host cluster with multi-component scheduling enabled on the `karmada-controller-manager`, `karmada-scheduler`, and `karmada-webhook` from the start — no additional patches or restarts needed.

Reviewer: The previous explanation of each component's role in `MultiplePodTemplatesScheduling` works well and can be retained.

RUN `karmadactl init --karmada-controller-manager-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true" --karmada-scheduler-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true" --karmada-webhook-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true"`{{exec}}

This sets up the Karmada control plane on the host cluster with multi-component scheduling enabled on the `karmada-controller-manager`, `karmada-scheduler`, and `karmada-webhook` from the start — no additional patches or restarts needed.

- **karmada-controller-manager**: Parses the workload using the Resource Interpreter framework and populates the `spec.components` array in the `ResourceBinding` to declare the resource requests of all sub-components.
- **karmada-scheduler**: Reads the `spec.components` array to calculate the total aggregated resources needed, ensuring the workload is only scheduled to member clusters with sufficient capacity to co-locate all components.

> - **karmada-webhook**: Intercepts the scheduling policies and validates the multi-component configurations.

Reviewer: ```suggestion
- **karmada-webhook**: Validates the multi-component fields within incoming `ResourceBinding` objects.
```

### Initialize Karmada control plane

**Initialize Karmada control plane:**

RUN `karmadactl init`{{exec}}

This sets up the Karmada control plane on the host cluster, including API server and controllers.

**Verify initialization:**

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config config get-contexts karmada-apiserver`{{exec}}

This outputs the `karmada-apiserver` context, ensuring that it is available and configured correctly.

### Enable the Multi-Component Scheduling feature gate

Multi-component scheduling (`MultiplePodTemplatesScheduling`) is currently an **Alpha** feature in Karmada and is **disabled by default**. We need to explicitly enable it on the `karmada-controller-manager`, `karmada-scheduler`, and `karmada-webhook` components.

> > **Note:** Because these components are running as native Pods on the underlying host cluster, we patch them using the default `kubectl` context, **not** the Karmada API server kubeconfig. We also temporarily change their deployment strategy to `Recreate` to prevent resource deadlocks during the rollout.

Reviewer: > We also temporarily change their deployment strategy to `Recreate` to prevent resource deadlocks during the rollout.

Can you elaborate more on this?

Author: I have added a bit more detail in the latest commit

Reviewer: Thanks for the explanation. However, these three components don't seem to have resource configurations declared. So there shouldn't be deadlocks during their restart, right?

**Verify initialization:**

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config config get-contexts karmada-apiserver`{{exec}}

This outputs the `karmada-apiserver` context, ensuring that it is available and configured correctly.

### Enable the Multi-Component Scheduling feature gate

Multi-component scheduling (`MultiplePodTemplatesScheduling`) is currently an **Alpha** feature in Karmada and is **disabled by default**. We need to explicitly enable it on three core control plane components to ensure the entire scheduling pipeline can process multi-component workloads:

- **`karmada-webhook`**: Needs the feature gate to successfully validate the multi-component fields within incoming `ResourceBinding` objects.
- **`karmada-controller-manager`**: Requires it to execute custom Resource Interpreters that extract the specific components, and to build the comprehensive `ResourceBinding` that contains them.
- **`karmada-scheduler`**: Uses it to obtain the detailed resource requirements of the workload, ensuring the selected target cluster has sufficient capacity to host the entire complex workload.

> > **Note:** Because these components are running as native Pods on the underlying host cluster, we patch them using the default `kubectl` context, **not** the Karmada API server kubeconfig. We temporarily change their deployment strategy to `Recreate`. This isn't strictly for resource limitations, but to prevent a race condition in this tutorial where the old leader pod processes incoming workloads (with the feature gate disabled) while the new pod is starting up.

Reviewer: > This isn't strictly for resource limitations, but to prevent a race condition in this tutorial where the old leader pod processes incoming workloads (with the feature gate disabled) while the new pod is starting up.

If you have concerns about this, you can set the corresponding feature gates via command-line arguments when running the command `karmadactl init --xxxx` to avoid restarts later. This will reduce complex operations and extra explanations down the line.

Run `karmadactl init --help` to check the usage.

Author: Thank you for the suggestion to use the karmadactl init flags.

I checked karmadactl init --help and have updated the initialization command in Step 3 to pass the feature gates directly at start time using:

--karmada-controller-manager-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true"
--karmada-scheduler-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true"
--karmada-webhook-extra-args="--feature-gates=MultiplePodTemplatesScheduling=true"

### Initialize Karmada control plane

**Initialize Karmada control plane:**

RUN `karmadactl init`{{exec}}

This sets up the Karmada control plane on the host cluster, including API server and controllers.

**Verify initialization:**

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config config get-contexts karmada-apiserver`{{exec}}

This outputs the `karmada-apiserver` context, ensuring that it is available and configured correctly.

### Enable the Multi-Component Scheduling feature gate

Multi-component scheduling (`MultiplePodTemplatesScheduling`) is currently an **Alpha** feature in Karmada and is **disabled by default**. We need to explicitly enable it on three core control plane components to ensure the entire scheduling pipeline can process multi-component workloads:

> - **`karmada-webhook`**: Needs the feature gate to successfully validate and mutate the multi-component fields within incoming `ResourceBinding` and `PropagationPolicy` objects.

Reviewer: ```suggestion
- **`karmada-webhook`**: Needs the feature gate to successfully validate the multi-component fields within incoming `ResourceBinding` objects.
```

- **`karmada-controller-manager`**: Requires it to execute custom Resource Interpreters that extract the specific components, and to build the comprehensive `ResourceBinding` that contains them.

> - **`karmada-scheduler`**: Uses it to compute the aggregate resource requirements of all extracted components, ensuring the selected target cluster has sufficient capacity to host the entire complex workload.

Reviewer: ```suggestion
- **`karmada-scheduler`**: Uses it to obtain the detailed resource requirements of the workload, ensuring the selected target cluster has sufficient capacity to host the entire complex workload.
```
